# HG changeset patch
# User Lachlan Andrew <lachlanbis@gmail.com>
# Date 1457953021 -39600
#      Mon Mar 14 21:57:01 2016 +1100
# Node ID 30a0756e6318bce309fe0d5be38de70b20d73351
# Parent  6e9f5408c0dbfa1aa1c104f1987db2d15ae6d657
Go to right editor line number, even with folding (bug #47414)

* file-editor-tab.cc (center_current_line, handle_context_menu_edit):
Expand current line, if it is folded.  Convert "document" line number
to "visible" line number before calling SCI_SETFIRSTVISIBLELINE.

diff --git a/libgui/src/m-editor/file-editor-tab.cc b/libgui/src/m-editor/file-editor-tab.cc
--- a/libgui/src/m-editor/file-editor-tab.cc
+++ b/libgui/src/m-editor/file-editor-tab.cc
@@ -309,17 +309,20 @@ file_editor_tab::handle_context_menu_edi
         break;
       if ((pos_fct = rxfun4.indexIn (lines.at (line))) != -1)
         break;
     }
 
   if (pos_fct > -1)
     { // reg expr. found: it is an internal function
       _edit_area->setCursorPosition (line, pos_fct);
-      _edit_area->SendScintilla (2613, line); // SCI_SETFIRSTVISIBLELINE
+      _edit_area->SendScintilla (2232, line);     // SCI_ENSUREVISIBLE
+                                                  // SCI_VISIBLEFROMDOCLINE
+      int vis_line = _edit_area->SendScintilla(2220, line);
+      _edit_area->SendScintilla (2613, vis_line); // SCI_SETFIRSTVISIBLELINE
       return;
     }
 
   emit edit_mfile_request (word_at_cursor, _file_name, _ced, -1);
 }
 
 // If "dbstop if ..." selected from context menu, create a conditional
 // breakpoint.  The default condition is (a) the existing condition if there
@@ -2469,23 +2472,30 @@ file_editor_tab::center_current_line (bo
 {
   long int visible_lines
     = _edit_area->SendScintilla (QsciScintillaBase::SCI_LINESONSCREEN);
 
   if (visible_lines > 2)
     {
       int line, index;
       _edit_area->getCursorPosition (&line, &index);
+      // compensate for "folding":
+      // step 1: expand the current line, if it was folded
+      _edit_area->SendScintilla (2232, line);   // SCI_ENSUREVISIBLE
+
+      // step 2: map file line num to "visible" one // SCI_VISIBLEFROMDOCLINE
+      int vis_line = _edit_area->SendScintilla (2220, line);
 
       int first_line = _edit_area->firstVisibleLine ();
 
-      if (always || line == first_line || line > first_line + visible_lines - 2)
+      if (always || vis_line == first_line
+          || vis_line > first_line + visible_lines - 2)
         {
-          first_line = first_line + (line - first_line - (visible_lines-1)/2);
-          _edit_area->SendScintilla (2613,first_line); // SCI_SETFIRSTVISIBLELINE
+          first_line += (vis_line - first_line - (visible_lines - 1) / 2);
+          _edit_area->SendScintilla (2613, first_line); // SCI_SETFIRSTVISIBLELINE
         }
     }
 }
 
 void
 file_editor_tab::handle_lines_changed ()
 {
   // the related signal is emitted before cursor-move-signal!
